libdill.h

```diff
@@ -967,6 +967,75 @@ DILL_EXPORT int dill_ws_response_key(
 #define ws_response_key dill_ws_response_key
 #endif
 
+/******************************************************************************/
+/*  SOCKS5                                                                    */
+/******************************************************************************/
+
+// SOCKS5 client commands
+#define DILL_SOCKS5_CONNECT (0x01)
+#define DILL_SOCKS5_BIND (0x02)
+#define DILL_SOCKS5_UDP_ASSOCIATE (0x03)
+
+// SOCKS5 server reply codes
+#define DILL_SOCKS5_SUCCESS (0x00)
+#define DILL_SOCKS5_GENERAL_FAILURE (0x01)
+#define DILL_SOCKS5_CONNECTION_NOT_ALLOWED (0x02)
+#define DILL_SOCKS5_NETWORK_UNREACHABLE (0x03)
+#define DILL_SOCKS5_HOST_UNREACHABLE (0x04)
+#define DILL_SOCKS5_CONNECTION_REFUSED (0x05)
+#define DILL_SOCKS5_TTL_EXPIRED (0x06)
+#define DILL_SOCKS5_COMMAND_NOT_SUPPORTED (0x07)
+#define DILL_SOCKS5_ADDRESS_TYPE_NOT_SUPPORTED (0x08)
+
+
+typedef int dill_socks5_auth_function(const char *username,
+    const char *password);
+
+DILL_EXPORT int dill_socks5_client_connect(
+    int s, const char *username, const char *password,
+    struct dill_ipaddr *addr, int64_t deadline);
+
+DILL_EXPORT int dill_socks5_client_connectbyname(
+    int s, const char *username, const char *password, const char *hostname,
+    int port, int64_t deadline);
+
+DILL_EXPORT int dill_socks5_proxy_auth(
+    int s, dill_socks5_auth_function *auth_fn, int64_t deadline);
+
+DILL_EXPORT int dill_socks5_proxy_recvcommand(
```

Comment:
How does proxy know whether to call recvcommand or recvcomandbyname?

Reply:
If the proxy wants to use DNS to resolve names then you can use recvcommand(), if the proxy has it's own application-specific name resolution (e.g. for tor .onion names) then you can use recvcommandbyname().

If it was required to only have one of these, recvcommandbyname() is the required functionality. You can subsequently resolve the name via ipaddr_remote().